### PR TITLE
Correctly querying wire IDs for `SetState` and `SetBasisState` in runtime

### LINKED ISSUE #1044: Recycling the same `device.so`  by different `quantum.set_state` causes a crash
The following mlir has multiple functions using set_state on the same device shared object file and it crashes:
```
module @circuit_pytest {
  func.func public @jit_circuit_pytest(%arg0: tensor<f64>, %arg1: tensor<3xf64>, %arg2: tensor<f64>) -> tensor<f64> attributes {llvm.emit_c_interface} {
    %cst = arith.constant dense<[(1.000000e+00,0.000000e+00), (0.000000e+00,0.000000e+00), (0.000000e+00,0.000000e+00), (0.000000e+00,0.000000e+00)]> : tensor<4xcomplex<f64>>
    %0 = call @circuit_pytest(%cst, %arg0, %arg1, %arg2) : (tensor<4xcomplex<f64>>, tensor<f64>, tensor<3xf64>, tensor<f64>) -> tensor<f64>
    return %0 : tensor<f64>
  }

  func.func @circuit_pytest_tape_2(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<3xf64>) -> tensor<f64> {
    quantum.device["/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", "LightningSimulator", "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    %3:2 = quantum.set_state(%arg0) %1, %2 : (tensor<4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
    %extracted_slice = tensor.extract_slice %arg1[2] [1] [1] : tensor<3xf64> to tensor<1xf64>
    %collapsed = tensor.collapse_shape %extracted_slice [] : tensor<1xf64> into tensor<f64>
    %extracted = tensor.extract %collapsed[] : tensor<f64>
    %out_qubits = quantum.custom "RX"(%extracted) %3#1 : !quantum.bit
    %4 = quantum.namedobs %3#0[ PauliZ] : !quantum.obs
    %5 = quantum.expval %4 : f64
    %from_elements = tensor.from_elements %5 : tensor<f64>
    %6 = quantum.insert %0[ 0], %3#0 : !quantum.reg, !quantum.bit
    %7 = quantum.insert %6[ 1], %out_qubits : !quantum.reg, !quantum.bit
    quantum.dealloc %7 : !quantum.reg
    quantum.device_release
    return %from_elements : tensor<f64>
  }

  func.func @circuit_pytest_tape_1(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<3xf64>) -> tensor<f64> {
    quantum.device["/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", "LightningSimulator", "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    %3:2 = quantum.set_state(%arg0) %1, %2 : (tensor<4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
    %4 = quantum.namedobs %3#0[ PauliZ] : !quantum.obs
    %5 = quantum.expval %4 : f64
    %from_elements = tensor.from_elements %5 : tensor<f64>
    %6 = quantum.insert %0[ 0], %3#0 : !quantum.reg, !quantum.bit
    %extracted_slice = tensor.extract_slice %arg1[1] [1] [1] : tensor<3xf64> to tensor<1xf64>
    %collapsed = tensor.collapse_shape %extracted_slice [] : tensor<1xf64> into tensor<f64>
    %extracted = tensor.extract %collapsed[] : tensor<f64>
    %out_qubits = quantum.custom "RX"(%extracted) %3#1 : !quantum.bit
    %7 = quantum.insert %6[ 1], %out_qubits : !quantum.reg, !quantum.bit
    quantum.dealloc %7 : !quantum.reg
    quantum.device_release
    return %from_elements : tensor<f64>
  }

  func.func @circuit_pytest_tape_0(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<3xf64>) -> tensor<f64> {
    quantum.device["/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", "LightningSimulator", "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    %3:2 = quantum.set_state(%arg0) %1, %2 : (tensor<4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
    %4 = quantum.namedobs %3#0[ PauliZ] : !quantum.obs
    %5 = quantum.expval %4 : f64
    %from_elements = tensor.from_elements %5 : tensor<f64>
    %6 = quantum.insert %0[ 0], %3#0 : !quantum.reg, !quantum.bit
    %extracted_slice = tensor.extract_slice %arg1[0] [1] [1] : tensor<3xf64> to tensor<1xf64>
    %collapsed = tensor.collapse_shape %extracted_slice [] : tensor<1xf64> into tensor<f64>
    %extracted = tensor.extract %collapsed[] : tensor<f64>
    %out_qubits = quantum.custom "RX"(%extracted) %3#1 : !quantum.bit
    %7 = quantum.insert %6[ 1], %out_qubits : !quantum.reg, !quantum.bit
    quantum.dealloc %7 : !quantum.reg
    quantum.device_release
    return %from_elements : tensor<f64>
  }

  func.func private @circuit_pytest(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<f64>, %arg2: tensor<3xf64>, %arg3: tensor<f64>) -> tensor<f64> attributes {diff_method = "parameter-shift", llvm.linkage = #llvm.linkage<internal>, qnode} {
    %0 = call @circuit_pytest_tape_0(%arg0, %arg2) : (tensor<4xcomplex<f64>>, tensor<3xf64>) -> tensor<f64>
    %1 = call @circuit_pytest_tape_1(%arg0, %arg2) : (tensor<4xcomplex<f64>>, tensor<3xf64>) -> tensor<f64>
    %2 = call @circuit_pytest_tape_2(%arg0, %arg2) : (tensor<4xcomplex<f64>>, tensor<3xf64>) -> tensor<f64>
    return %0 : tensor<f64>
  }

  func.func @setup() {
    quantum.init
    return
  }
  func.func @teardown() {
    quantum.finalize
    return
  }
}
```

```
malloc(): unaligned tcache chunk detected
```

However, if we do one of the two following, then there won't be a crash:
1. create copies of the device by `cp librtd_lightning.so librtd_lightning_1.so && cp librtd_lightning.so librtd_lightning_2.so`, and use these copies in  `quantum.device`
2. remove all the `quantum.set_state`, i.e. 
```
module @circuit_pytest {
  func.func public @jit_circuit_pytest(%arg0: tensor<f64>, %arg1: tensor<3xf64>, %arg2: tensor<f64>) -> tensor<f64> attributes {llvm.emit_c_interface} {
    %cst = arith.constant dense<[(1.000000e+00,0.000000e+00), (0.000000e+00,0.000000e+00), (0.000000e+00,0.000000e+00), (0.000000e+00,0.000000e+00)]> : tensor<4xcomplex<f64>>
    %0 = call @circuit_pytest(%cst, %arg0, %arg1, %arg2) : (tensor<4xcomplex<f64>>, tensor<f64>, tensor<3xf64>, tensor<f64>) -> tensor<f64>
    return %0 : tensor<f64>
  }

  func.func @circuit_pytest_tape_2(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<3xf64>) -> tensor<f64> {
    quantum.device["/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", "LightningSimulator", "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    //%3:2 = quantum.set_state(%arg0) %1, %2 : (tensor<4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
    %extracted_slice = tensor.extract_slice %arg1[2] [1] [1] : tensor<3xf64> to tensor<1xf64>
    %collapsed = tensor.collapse_shape %extracted_slice [] : tensor<1xf64> into tensor<f64>
    %extracted = tensor.extract %collapsed[] : tensor<f64>
    //%out_qubits = quantum.custom "RX"(%extracted) %3#1 : !quantum.bit
    %out_qubits = quantum.custom "RX"(%extracted) %2 : !quantum.bit
    //%4 = quantum.namedobs %3#0[ PauliZ] : !quantum.obs
    %4 = quantum.namedobs %1[ PauliZ] : !quantum.obs
    %5 = quantum.expval %4 : f64
    %from_elements = tensor.from_elements %5 : tensor<f64>
    //%6 = quantum.insert %0[ 0], %3#0 : !quantum.reg, !quantum.bit
    %6 = quantum.insert %0[ 0], %1 : !quantum.reg, !quantum.bit
    %7 = quantum.insert %6[ 1], %out_qubits : !quantum.reg, !quantum.bit
    quantum.dealloc %7 : !quantum.reg
    quantum.device_release
    return %from_elements : tensor<f64>
  }

  func.func @circuit_pytest_tape_1(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<3xf64>) -> tensor<f64> {
    quantum.device["/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", "LightningSimulator", "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    //%3:2 = quantum.set_state(%arg0) %1, %2 : (tensor<4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
    //%4 = quantum.namedobs %3#0[ PauliZ] : !quantum.obs
    %4 = quantum.namedobs %1[ PauliZ] : !quantum.obs
    %5 = quantum.expval %4 : f64
    %from_elements = tensor.from_elements %5 : tensor<f64>
    //%6 = quantum.insert %0[ 0], %3#0 : !quantum.reg, !quantum.bit
    %6 = quantum.insert %0[ 0], %1 : !quantum.reg, !quantum.bit
    %extracted_slice = tensor.extract_slice %arg1[1] [1] [1] : tensor<3xf64> to tensor<1xf64>
    %collapsed = tensor.collapse_shape %extracted_slice [] : tensor<1xf64> into tensor<f64>
    %extracted = tensor.extract %collapsed[] : tensor<f64>
    //%out_qubits = quantum.custom "RX"(%extracted) %3#1 : !quantum.bit
    %out_qubits = quantum.custom "RX"(%extracted) %2 : !quantum.bit
    %7 = quantum.insert %6[ 1], %out_qubits : !quantum.reg, !quantum.bit
    quantum.dealloc %7 : !quantum.reg
    quantum.device_release
    return %from_elements : tensor<f64>
  }

  func.func @circuit_pytest_tape_0(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<3xf64>) -> tensor<f64> {
    quantum.device["/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", "LightningSimulator", "{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}"]
    %0 = quantum.alloc( 2) : !quantum.reg
    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
    //%3:2 = quantum.set_state(%arg0) %1, %2 : (tensor<4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
    //%4 = quantum.namedobs %3#0[ PauliZ] : !quantum.obs
    %4 = quantum.namedobs %1[ PauliZ] : !quantum.obs
    %5 = quantum.expval %4 : f64
    %from_elements = tensor.from_elements %5 : tensor<f64>
    //%6 = quantum.insert %0[ 0], %3#0 : !quantum.reg, !quantum.bit
    %6 = quantum.insert %0[ 0], %1 : !quantum.reg, !quantum.bit
    %extracted_slice = tensor.extract_slice %arg1[0] [1] [1] : tensor<3xf64> to tensor<1xf64>
    %collapsed = tensor.collapse_shape %extracted_slice [] : tensor<1xf64> into tensor<f64>
    %extracted = tensor.extract %collapsed[] : tensor<f64>
    //%out_qubits = quantum.custom "RX"(%extracted) %3#1 : !quantum.bit
    %out_qubits = quantum.custom "RX"(%extracted) %2 : !quantum.bit
    %7 = quantum.insert %6[ 1], %out_qubits : !quantum.reg, !quantum.bit
    quantum.dealloc %7 : !quantum.reg
    quantum.device_release
    return %from_elements : tensor<f64>
  }

  func.func private @circuit_pytest(%arg0: tensor<4xcomplex<f64>>, %arg1: tensor<f64>, %arg2: tensor<3xf64>, %arg3: tensor<f64>) -> tensor<f64> attributes {diff_method = "parameter-shift", llvm.linkage = #llvm.linkage<internal>, qnode} {
    %0 = call @circuit_pytest_tape_0(%arg0, %arg2) : (tensor<4xcomplex<f64>>, tensor<3xf64>) -> tensor<f64>
    %1 = call @circuit_pytest_tape_1(%arg0, %arg2) : (tensor<4xcomplex<f64>>, tensor<3xf64>) -> tensor<f64>
    %2 = call @circuit_pytest_tape_2(%arg0, %arg2) : (tensor<4xcomplex<f64>>, tensor<3xf64>) -> tensor<f64>
    return %0 : tensor<f64>
  }
  func.func @setup() {
    quantum.init
    return
  }
  func.func @teardown() {
    quantum.finalize
    return
  }
}
```

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**
Recycling a device when a previous execution involves stateprep causes a crash. See #1044 

**Description of the Change:**
In `LightningSimulator`, `SetState` and `SetBasisState` now correctly query the DevQubits from the SimQubits in its qubit_manger's map.

**Benefits:** We can now have multiple qnode functions involving stateprep in a workflow

**Related GitHub Issues:** closes #1044 

[sc-71894]

### CODE REVIEW

**Comment by dime10** [comment]
Should we not look into why it crashes in the first place?

**Comment by paul0403** [comment]
> Should we not look into why it crashes in the first place?

Yes, draft 1 is trying to contain this issue within Catalyst but I agree the root cause needs to be addressed in Lightning. Chatted with @maliasadi regarding how the two repos see each other (also tagging to keep Ali in the loop on this issue).

**Comment by paul0403** [comment]
Root cause: not properly resetting the wires' `va_list` in `__catalyst__qis__SetState`: https://github.com/PennyLaneAI/catalyst/blob/1950dd1323a1267396dfe59640d2e119f8ae21f7/runtime/lib/capi/RuntimeCAPI.cpp#L467

e.g. for a 2-wire circuit, the first execution will make this `wires` vector {0,1}, but the second execution makes it {2,3}, so causing "index out of range" style segfaults downstream in the backend, specifically https://github.com/PennyLaneAI/pennylane-lightning/blob/97554407e114db41690b665c01f342c8ec27b98a/pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp#L809, where the iterator to `std::vector::erase(it)` is out of bounds.

**Comment by paul0403** [comment]
A supernatural feeling when it's Friday 6pm and you spent a whole day on a bug and the fix is one line

File: runtime/lib/backend/common/QubitManager.hpp [hunk 1]
```diff
@@ -141,6 +141,7 @@ class QubitManager {
     {
         // Release all qubits by clearing the map.
         this->qubits_map.clear();
+        this->next_idx = 0;
```
Comment (paul0403) [inline runtime/lib/backend/common/QubitManager.hpp:144]: self changelog reminder

**Comment by paul0403** [comment]
This bug came from mismatched indices. 

Consider the following IR we send to the runtime capi (compiled from the added test example):
```
define internal { ptr, ptr, i64, [1 x i64], [1 x i64] } @f(ptr %0, ptr %1, i64 %2, i64 %3, i64 %4) {
  %6 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } undef, ptr %0, 0
  %7 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %6, ptr %1, 1
  %8 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %7, i64 %2, 2
  %9 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %8, i64 %3, 3, 0
  %10 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %9, i64 %4, 4, 0
  call void @__catalyst__rt__device_init(ptr @"/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", ptr @LightningSimulator, ptr @"{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}")
  %11 = call ptr @__catalyst__rt__qubit_allocate_array(i64 2)
  %12 = call ptr @__catalyst__rt__array_get_element_ptr_1d(ptr %11, i64 0)
  %13 = load ptr, ptr %12, align 8
  %14 = call ptr @__catalyst__rt__array_get_element_ptr_1d(ptr %11, i64 1)
  %15 = load ptr, ptr %14, align 8
  %16 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, i64 1, align 8
  store { ptr, ptr, i64, [1 x i64], [1 x i64] } %10, ptr %16, align 8
  call void (ptr, i64, ...) @__catalyst__qis__SetState(ptr %16, i64 2, ptr %13, ptr %15)
  %17 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (double, ptr null, i32 4) to i64))
  %18 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } undef, ptr %17, 0
  %19 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %18, ptr %17, 1
  %20 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %19, i64 0, 2
  %21 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %20, i64 4, 3, 0
  %22 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %21, i64 1, 4, 0
  %23 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, i64 1, align 8
  store { ptr, ptr, i64, [1 x i64], [1 x i64] } %22, ptr %23, align 8
  call void (ptr, i64, ...) @__catalyst__qis__Probs(ptr %23, i64 2, ptr %13, ptr %15)
  call void @__catalyst__rt__qubit_release_array(ptr %11)
  call void @__catalyst__rt__device_release()
  ret { ptr, ptr, i64, [1 x i64], [1 x i64] } %22
}

define internal { ptr, ptr, i64, [1 x i64], [1 x i64] } @g(ptr %0, ptr %1, i64 %2, i64 %3, i64 %4) {
  %6 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } undef, ptr %0, 0
  %7 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %6, ptr %1, 1
  %8 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %7, i64 %2, 2
  %9 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %8, i64 %3, 3, 0
  %10 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %9, i64 %4, 4, 0
  call void @__catalyst__rt__device_init(ptr @"/home/paul.wang/catalyst_new/catalyst/frontend/catalyst/utils/../../../runtime/build/lib/librtd_lightning.so", ptr @LightningSimulator, ptr @"{'shots': 0, 'mcmc': False, 'num_burnin': 0, 'kernel_name': None}")
  %11 = call ptr @__catalyst__rt__qubit_allocate_array(i64 2)
  %12 = call ptr @__catalyst__rt__array_get_element_ptr_1d(ptr %11, i64 0)
  %13 = load ptr, ptr %12, align 8
  %14 = call ptr @__catalyst__rt__array_get_element_ptr_1d(ptr %11, i64 1)
  %15 = load ptr, ptr %14, align 8
  %16 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, i64 1, align 8
  store { ptr, ptr, i64, [1 x i64], [1 x i64] } %10, ptr %16, align 8
  call void (ptr, i64, ...) @__catalyst__qis__SetState(ptr %16, i64 2, ptr %13, ptr %15)
  %17 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (double, ptr null, i32 4) to i64))
  %18 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } undef, ptr %17, 0
  %19 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %18, ptr %17, 1
  %20 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %19, i64 0, 2
  %21 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %20, i64 4, 3, 0
  %22 = insertvalue { ptr, ptr, i64, [1 x i64], [1 x i64] } %21, i64 1, 4, 0
  %23 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, i64 1, align 8
  store { ptr, ptr, i64, [1 x i64], [1 x i64] } %22, ptr %23, align 8
  call void (ptr, i64, ...) @__catalyst__qis__Probs(ptr %23, i64 2, ptr %13, ptr %15)
  call void @__catalyst__rt__qubit_release_array(ptr %11)
  call void @__catalyst__rt__device_release()
  ret { ptr, ptr, i64, [1 x i64], [1 x i64] } %22
}
```

The first call, `__catalyst__rt__qubit_allocate_array`, calls the `AllocateQubits` of the quantum device objects and (more or less) returns whatever `AllocateQubits` returns (https://github.com/PennyLaneAI/catalyst/blob/main/runtime/lib/capi/RuntimeCAPI.cpp#L325). The quantum device's `AllocateQubits` calls the device's `QubitManager`'s `AllocateRange` (https://github.com/PennyLaneAI/catalyst/blob/main/runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp#L39) and returns its results. 

The `AllocateRange` will return [0,1] for the first call by `f` and [2,3] for the second call by `g` (which gets stored into the SSA value `%11` in the above example), because it collects the `QubitManager`'s `next_idx`, which does not get reset to zero when the manager is released (https://github.com/PennyLaneAI/catalyst/blob/main/runtime/lib/backend/common/QubitManager.hpp#L129). 

Then, when `__catalyst__qis__SetState` is called with arguments `%13, %15` ([0,1] for the first call, [2,3] for the second call), it passes these values down to the device's `SetState` (https://github.com/PennyLaneAI/catalyst/blob/main/runtime/lib/capi/RuntimeCAPI.cpp#L467). The device then uses these indices to try to set the statevector (https://github.com/PennyLaneAI/catalyst/blob/main/runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp#L112), leading to "vector-index-out-of-range" segfaults, specifically at https://github.com/PennyLaneAI/pennylane-lightning/blob/97554407e114db41690b665c01f342c8ec27b98a/pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp#L809

**Comment by paul0403** [comment]
The root cause is because `SetState` (and `BasisState`) did not query the qubit_manager map values and just used the keys.

**Review by erick-xanadu — APPROVED** [review]
This makes sense! Thanks! I must have forgotten to add the `getDeviceWires` function call. All good on my part.

**Comment by paul0403** [comment]
I guess we need to update the kokkos ones in lightning as well? https://github.com/PennyLaneAI/pennylane-lightning/blob/4c6e4bb750b7d976fe661677895692ef6c129723/pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosSimulator.cpp#L140

Interestingly, the failure in #1044 does **not** happen on kokkos.

File: runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp [hunk 2]
```diff
@@ -108,14 +108,16 @@ void LightningSimulator::SetState(DataView<std::complex<double>, 1> &data,
                                   std::vector<QubitIdType> &wires)
 {
     std::vector<std::complex<double>> data_vector(data.begin(), data.end());
-    std::vector<std::size_t> wires_size_t(wires.begin(), wires.end());
+    auto &&dev_wires = getDeviceWires(wires);
+    std::vector<std::size_t> wires_size_t(dev_wires.begin(), dev_wires.end());
```
Comment (dime10) [inline runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp:112]: I don't think we need to repackage the wires into an additional vector

Comment (paul0403) [reply runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp:112]: This was suggested on the lightning side too (https://github.com/PennyLaneAI/pennylane-lightning/pull/869), I will change this

**Pull request merged**